### PR TITLE
Implement a class using posix timer

### PR DESCRIPTION
**Pull request by udzura**
Hi, I have added POSIX timer-based API, using `timer_create(2)` family.

TODOs are:

* [x] Follow the interfaces of [TimerThread](https://github.com/matsumotory/mruby-timer-thread/pull/1/files#diff-324b691c1cada8f8510e81449b451390R20)
* [x] Resolve conflict

### CODE REVIEW

**Comment by udzura** [comment]
All done! Usage is as test case: https://github.com/matsumotory/mruby-timer-thread/pull/2/files#diff-50fda337a4df58a689cafe63edf37c9e

File: test/posix_test.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    count = 0
+
+    pt = Timer::POSIX.new(signal: :SIGRT1) do
+      count += 1
+    end
+    start = Time.now.to_i * 1000 + Time.now.usec / 1000
+    pt.run timer_msec, timer_msec
+
+    while count < 4 do
+      usleep 1000
```
Comment (matsumotory) [inline test/posix_test.rb]: why counter under 4？

File: test/posix_test.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  assert("Timer::POSIX with RTSignal, interval timer and block") do
+    timer_msec = 200
+    count = 0
+
+    pt = Timer::POSIX.new(signal: :SIGRT1) do
+      count += 1
+    end
+    start = Time.now.to_i * 1000 + Time.now.usec / 1000
+    pt.run timer_msec, timer_msec
```
Comment (matsumotory) [inline test/posix_test.rb]: timer_msec args is duplicated?

File: src/timer_thread.c [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+{
+  mrb_timer_posix_data *data;
+  timer_t *timer_ptr;
+  mrb_value options = mrb_nil_value();
+  mrb_value blk = mrb_nil_value();
+
+  struct sigevent sev;
+  sev.sigev_signo = -1;
+
+  struct sigaction sact;
```
Comment (matsumotory) [inline src/timer_thread.c]: I'd like to support C90.

File: src/timer_thread.c [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+static mrb_value mrb_timer_posix_status_raw(mrb_state *mrb, mrb_value self)
+{
+  mrb_timer_posix_data *data = DATA_PTR(self);
+  struct itimerspec ts;
+
+  if (timer_gettime(*(data->timer_ptr), &ts) == -1) {
+    mrb_sys_fail(mrb, "timer_gettime");
+  }
+
+  mrb_value ret = mrb_hash_new_capa(mrb, 2);
```
Comment (matsumotory) [inline src/timer_thread.c]: I'd like to support C90.

File: src/timer_thread.c [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+  mrb_free(mrb, data);
+}
+
+static const struct mrb_data_type mrb_timer_posix_data_type = {"mrb_timer_posix_data", mrb_timer_posix_free};
+
+static mrb_value mrb_rtsignal_get(mrb_state *mrb, mrb_value self)
+{
+  mrb_int idx;
+  if (mrb_get_args(mrb, "i", &idx) == -1) {
+    mrb_sys_fail(mrb, "mrb_get_args");
```
Comment (matsumotory) [inline src/timer_thread.c]: You should use raise when isn't syscall error.

File: src/timer_thread.c [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+  if (SIGRTMIN + (int)idx > SIGRTMAX) {
+    mrb_raise(mrb, E_ARGUMENT_ERROR, "RTSignal indx too large");
+  }
+
+  return mrb_fixnum_value(SIGRTMIN + (int)idx);
+}
+
+static mrb_sym mrb_rtsignal_iv_key(mrb_state *mrb, int signo)
+{
+  char key[sizeof("__mrb_timer_trap_proc_") + 3];
```
Comment (matsumotory) [inline src/timer_thread.c]: I'd like to support C90.

**Review by matsumotory — CHANGES_REQUESTED** [review]

**Comment by matsumotory** [comment]
Also check that you can run multiple timers.

File: test/posix_test.rb (hunk 1 above)
Comment (udzura) [reply test/posix_test.rb]: `pt.run timer_msec, timer_msec` 's second arg is interval(like libuv's timer).

This checks the count will be incremented 4 times (first invoke + interval invoke x 3)
via L28 - 30's block.

**Comment by udzura** [comment]
> multiple timers

This means 2 >= timers in the same code? Intervel timer is tested via https://github.com/matsumotory/mruby-timer-thread/pull/2/files#r108030011

File: src/timer_thread.c (hunk 4 above)
Comment (udzura) [reply src/timer_thread.c]: yeah

File: src/timer_thread.c (hunk 5 above)
Comment (udzura) [reply src/timer_thread.c]: I see!

File: test/posix_test.rb (hunk 1 above)
Comment (matsumotory) [reply test/posix_test.rb]: Thanks, I understood. Please add the docs to readme.

**Comment by matsumotory** [comment]
multiple timer means have multiple Timer::Posix instance in the same code. timer_create sometimes was used for such purposes.

I think top self have realtime signal key as instance variable is data race.

**Comment by udzura** [comment]
> top self have realtime signal key as instance variable is data race

That sounds reasonable.

I want to remove handler code block support in Timer::POSIX class itself and delegete block registration to [mruby-signal-thread](https://github.com/pyama86/mruby-signal-thread/pull/3). How about this?

**Comment by udzura** [comment]
>  run multiple timers

Done via https://github.com/matsumotory/mruby-timer-thread/pull/2/commits/865a5a33ae0f8f19372f3fb2e68d580331a3b0c1

File: Rakefile [hunk 7]
```diff
@@ -10,6 +10,11 @@ task :compile => :mruby do
   sh "cd mruby && rake all MRUBY_CONFIG=\"#{MRUBY_CONFIG}\""
 end
 
+desc "memtest"
+task :memtest => :compile do
+  sh %q<valgrind ./mruby/bin/mruby -e '100.times {t =  Timer::POSIX.new ; p t}'>
+end
```
Comment (matsumotory) [inline Rakefile]: Wow, very useful!

File: src/timer_thread.c [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+  case MRB_TT_FIXNUM:
+    sig = mrb_fixnum(vsig);
+    if (sig < 0 || sig >= SIGRTMAX) {
+      mrb_raisef(mrb, E_ARGUMENT_ERROR, "invalid signal number (%S)", vsig);
+    }
+    break;
+  case MRB_TT_SYMBOL:
+    s = mrb_sym2name(mrb, mrb_symbol(vsig));
+    if (!s)
+      mrb_raise(mrb, E_ARGUMENT_ERROR, "bad signal");
```
Comment (matsumotory) [inline src/timer_thread.c]: It is better to unify it to either `if (hoge) fuga;` or `if (hoge) { fuga; }` notation.

File: src/timer_thread.c [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+}
+
+static const struct mrb_data_type mrb_timer_posix_data_type = {"mrb_timer_posix_data", mrb_timer_posix_free};
+
+static mrb_value mrb_rtsignal_get(mrb_state *mrb, mrb_value self)
+{
+  mrb_int idx;
+  if (mrb_get_args(mrb, "i", &idx) == -1) {
+    mrb_raise(mrb, E_RUNTIME_ERROR, "Cannot get arguments");
+  }
```
Comment (matsumotory) [inline src/timer_thread.c:245]: It is better to unify it to either `if (hoge) fuga;` or `if (hoge) { fuga; }` notation.

**Review by matsumotory — CHANGES_REQUESTED** [review]
I pointed out a few things, but once these are resolved I will merge!

**Comment by matsumotory** [comment]
@udzura  Thank you for your great works! I added you to collaborators! Please check the invited e-mail.

**Comment by udzura** [comment]
Thanks! I got invitation 💪 

I'm going to fix `if`'s unification!

**Review by matsumotory — APPROVED** [review]

**Comment by matsumotory** [comment]
Thanks!

**Pull request merged**